mplc/multi_partner_learning.py

```diff
@@ -5,804 +5,478 @@
-                    self.log_collaborative_round_partner_result(partner, partner_index, model_evaluation_val_data)
-
-                    # Update the partner's model in the models' list
-                    self.save_model_for_partner(partner_model, partner_index)
-
-                    # Update iterative results
-                    self.update_iterative_results(partner_index, history)
-
-                partners_models = self.init_with_agg_models()
-                self.minibatch_index += 1
-
-            # At the end of each epoch, evaluate the model for early stopping on a global validation set
-            model_to_evaluate = partners_models[0]
-            model_evaluation_val_data = self.evaluate_model(model_to_evaluate, self.val_data)
-
-            current_val_loss = model_evaluation_val_data[0]
-            current_val_metric = model_evaluation_val_data[1]
-
-            self.score_matrix_collective_models[self.epoch_index, self.minibatch_count] = current_val_metric
-            self.loss_collective_models.append(current_val_loss)
-
-            logger.info(f"   Model evaluation at the end of the epoch: "
-                        f"{['%.3f' % elem for elem in model_evaluation_val_data]}")
 
-            self.epoch_index += 1
+# Supported multipartner learning approaches
 
-            logger.debug("      Checking if early stopping criteria are met:")
-            if self.is_early_stopping:
-                # Early stopping parameters
-                if (
-                        self.epoch_index >= constants.PATIENCE
-                        and current_val_loss > self.loss_collective_models[-constants.PATIENCE]
-                ):
-                    logger.debug("         -> Early stopping criteria are met, stopping here.")
-                    break
-                else:
-                    logger.debug("         -> Early stopping criteria are not met, continuing with training.")
-
-        logger.info("### Evaluating model on test data:")
-        model_evaluation_test_data = self.evaluate_model(model_to_evaluate, self.test_data)
-        logger.info(f"   Model metrics names: {model_to_evaluate.metrics_names}")
-        logger.info(f"   Model metrics values: {['%.3f' % elem for elem in model_evaluation_test_data]}")
-        self.test_score = model_evaluation_test_data[1]  # 0 is for the loss
-        self.nb_epochs_done = self.epoch_index
-
-        # Plot training history
-        if self.is_save_data:
-            self.save_data()
-
-        self.save_final_model_weights(model_to_evaluate)
-
-        logger.info("Training and evaluation on multiple partners: done.")
-        end = timer()
-        self.learning_computation_time = end - start
+MULTI_PARTNER_LEARNING_APPROACHES = {
+    "fedavg": FederatedAverageLearning,
+    "seq-pure": SequentialLearning,
+    "seq-with-final-agg": SequentialWithFinalAggLearning,
+    "seqavg": SequentialAverageLearning
+}
```

Comment:
Why don't you add `MplLabeLFlip` in there?
(For `SinglePartnerLearning` I suppose it's ok... 😄 )

Reply:
Ah it is a good question. I thought about it as a contributivity method, not as a mpl one. But it can be interesting to add it yes ! I wonder what append if we compute Shapley with Lflip